File: server/test/post_validation.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { postSchema, postPhotosSchema } from '../src/post_validation.js';

const valid = {
  category: 'Scrap', title: 'Metal for sale', description: 'Good quality scrap metal',
  price: '50', unit: 'kg', storeNumber: '0101',
};

test('every displayed post field is required', () => {
  assert.equal(postSchema.safeParse(valid).success, true);
  for (const key of Object.keys(valid)) {
    for (const value of ['', '   ', undefined]) {
      assert.equal(postSchema.safeParse({ ...valid, [key]: value }).success, false, key);
    }
  }
});

test('salary posts require salary but do not require a hidden unit', () => {
  for (const category of ['Need Worker', 'Need Job']) {
    assert.equal(postSchema.safeParse({ ...valid, category, unit: '' }).success, true);
    assert.equal(postSchema.safeParse({ ...valid, category, price: '' }).success, false);
  }
});

test('rejects invalid prices and units and requires one to three photos', () => {
  for (const price of [null, false, -1, 'NaN', 'Infinity', 'abc', 10000000000]) {
    assert.equal(postSchema.safeParse({ ...valid, price }).success, false);
  }
  assert.equal(postSchema.safeParse({ ...valid, unit: 'item' }).success, false);
  assert.equal(postPhotosSchema.safeParse([]).success, false);
  assert.equal(postPhotosSchema.safeParse([{}]).success, true);
  assert.equal(postPhotosSchema.safeParse([{}, {}, {}]).success, true);
  assert.equal(postPhotosSchema.safeParse([{}, {}, {}, {}]).success, false);
});

File: client/test/post_validation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:scrap_market/post_validation.dart';

void main() {
  String? validate({String? category = 'Scrap', String title = 'Metal for sale',
    String description = 'Good quality scrap metal', String price = '50',
    String? unit = 'kg', String storeNumber = '0101', int photoCount = 1}) =>
      validatePostFields(category: category, title: title, description: description,
        price: price, unit: unit, storeNumber: storeNumber, photoCount: photoCount);

  test('requires all displayed fields and at least one photo', () {
    expect(validate(), isNull);
    expect(validate(category: null), isNotNull);
    expect(validate(title: ''), isNotNull);
    expect(validate(description: ''), isNotNull);
    expect(validate(price: ''), isNotNull);
    expect(validate(unit: null), isNotNull);
    expect(validate(storeNumber: ''), isNotNull);
    expect(validate(photoCount: 0), isNotNull);
  });

  test('requires salary but skips units for employment posts', () {
    for (final category in ['Need Job', 'Need Worker']) {
      expect(validate(category: category, unit: null), isNull);
      expect(validate(category: category, price: '', unit: null), isNotNull);
    }
  });

  test('rejects invalid amounts and units', () {
    for (final price in [' ', '-1', 'NaN', 'Infinity', 'abc', '10000000000']) {
      expect(validate(price: price), isNotNull);
    }
    expect(validate(unit: 'item'), isNotNull);
  });
}
